pkg/ha/ha_parser.go

```diff
@@ -28,31 +28,34 @@ func (h *haParser) ParseData(tts []prompb.TimeSeries) (map[string][]model.Sample
 	dataSamples := make(map[string][]model.SamplesInfo)
 	rows := 0
 
-	var replicaName, clusterName string
-	if len(tts) > 0 {
-		s, _, err := model.LabelProtosToLabels(tts[0].Labels)
-		if err != nil {
-			return nil, rows, err
-		}
-		replicaName = s.GetReplicaName()
-		clusterName = s.GetClusterName()
+	if len(tts) == 0 {
+		return dataSamples, rows, nil
+	}
+
+	s, _, err := model.LabelProtosToLabels(tts[0].Labels)
+	if err != nil {
+		return nil, rows, err
 	}
+	replicaName := s.GetReplicaName()
+	clusterName := s.GetClusterName()
 
 	if err := checkClusterAndReplicaLabelAreSet(clusterName, replicaName); err != nil {
 		return nil, rows, err
 	}
 
 	// find samples time range
 	var minTUnix, maxTUnix int64
+	minTWasSet := false
 	for i := range tts {
 		t := &tts[i]
 		if len(t.Samples) == 0 {
 			continue
 		}
 
 		for _, sample := range t.Samples {
-			if sample.Timestamp < minTUnix || minTUnix == 0 {
+			if sample.Timestamp < minTUnix || !minTWasSet {
 				minTUnix = sample.Timestamp
+				minTWasSet = true
 			}
```

Comment:
Ooh that would be super great if samples are ordered 👍 and if that is defined by the protocol then we can even enforce it for non-prometheus inserts. I also hate the second iteration

Reply:
But enforcing time-ordered samples will limit us from back-filling right? I think for backfilling we can recommend using Promscale in normal mode instead of HA. But this enforcement will be a limitation in HA mode.